consensus/polybft/checkpoint_manager.go

```diff
@@ -0,0 +1,209 @@
+
+	// frequency at which checkpoints are sent to the rootchain
+	checkpointTimeInterval = 30 * time.Minute
+)
+
+// checkpointManager encapsulates logic for checkpoint data submission
+type checkpointManager struct {
+	// sender address
+	sender types.Address
+	// blockchain is abstraction for blockchain
+	blockchain blockchainBackend
+	// consensusBackend is abstraction for polybft consensus specific functions
+	consensusBackend polybftBackend
+	// rootchain represents abstraction for rootchain interaction
+	rootchain rootchainInteractor
+	// checkpointsOffset represents offset between checkpoint blocks (applicable only for non-epoch ending blocks)
+	checkpointsOffset uint64
+}
+
+// newCheckpointManager creates a new instance of checkpointManager
+func newCheckpointManager(sender types.Address, blockTime time.Duration, interactor rootchainInteractor,
+	blockchain blockchainBackend, backend polybftBackend) *checkpointManager {
+	r := interactor
+	if interactor == nil {
+		r = &defaultRootchainInteractor{}
+	}
+
+	return &checkpointManager{
+		sender:            sender,
+		blockchain:        blockchain,
+		consensusBackend:  backend,
+		rootchain:         r,
+		checkpointsOffset: uint64(checkpointTimeInterval.Milliseconds() / blockTime.Milliseconds()),
+	}
+}
+
+// getCurrentCheckpointID queries CheckpointManager smart contract and retrieves current checkpoint id
+func (c checkpointManager) getCurrentCheckpointID(epochNumber uint64) (uint64, error) {
+	checkpointIDMethodEncoded, err := currentCheckpointIDMethod.Encode([]interface{}{})
+	if err != nil {
+		return 0, fmt.Errorf("failed to encode currentCheckpointId function parameters for epoch=%d: %w",
+			epochNumber, err)
+	}
+
+	currentCheckpointID, err := c.rootchain.Call(c.sender, helper.CheckpointManagerAddress, checkpointIDMethodEncoded)
+	if err != nil {
+		return 0, fmt.Errorf("failed to invoke currentCheckpointId function on the rootchain for epoch=%d: %w",
+			epochNumber, err)
+	}
+
+	checkpointID, err := strconv.ParseUint(currentCheckpointID, 0, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to convert current checkpoint id '%s' to number for epoch=%d: %w",
+			currentCheckpointID, epochNumber, err)
+	}
+
+	return checkpointID, nil
+}
+
+// submitCheckpoint sends a transaction which with checkpoint data to the rootchain
+func (c checkpointManager) submitCheckpoint(latestHeader types.Header, epochNumber uint64) error {
+	checkpointID, err := c.getCurrentCheckpointID(epochNumber)
+	if err != nil {
+		return err
+	}
+
+	nonce, err := c.rootchain.GetPendingNonce(c.sender)
+	if err != nil {
+		return err
+	}
+
+	checkpointManagerAddr := ethgo.Address(helper.CheckpointManagerAddress)
+	txn := &ethgo.Transaction{
+		To: &checkpointManagerAddr,
+	}
+
+	// detect any pending (previously failed) checkpoints and send them
+	for blockNum := checkpointID + 1; blockNum < latestHeader.Number; blockNum++ {
```

Comment:
hmm. Do we need to handle that eth node of a validator got stuck? We can get dozens of the same checkpoint events.

Reply:
This for loop basically sends all the epoch ending blocks which were failed at some of the previous occasions. 

> We can get dozens of the same checkpoint events.

To me, `checkpointID` is the last checkpointed block. It is obtained from the smart contract (https://github.com/0xPolygon/polygon-edge/pull/861/files#diff-fce12e7457acbf9d31265d9e0f71a4340b87f09cb3dd52e2f61df69d0bbc142dR70). That should prevent sending same checkpoint data multiple times.

If you disagree, can you elaborate more?